pkg/api/helpers/errors.go

```diff
@@ -0,0 +1,66 @@
+package helpers
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"reflect"
+
+	"github.com/treeverse/lakefs/pkg/api"
+)
+
+var (
+	// ErrUnsupportedProtocol is the error returned when lakeFS server requests the client
+	// use a protocol it does not know about.  Until recompiled to use a newer client,
+	// upload the object using through the lakeFS server.
+	ErrUnsupportedProtocol = errors.New("unsupported protocol")
+
+	// ErrRequestFailed is an error returned for failing lakeFS server replies.
+	ErrRequestFailed = errors.New("request failed")
+)
+
+const minHTTPErrorStatusCode = 400
+
+// isOK returns true if statusCode is an OK HTTP status code: 0-399.
+func isOK(statusCode int) bool {
+	return statusCode < minHTTPErrorStatusCode
+}
+
+// ResponseAsError returns a ErrRequestFailed wrapping a response from the server.  It
+// searches for a non-nil unsuccessful HTTPResponse field and uses its message, along with a
+// Body that it assumes is an api.Error.
+func ResponseAsError(response interface{}) error {
+	r := reflect.Indirect(reflect.ValueOf(response))
+	if !r.IsValid() {
+		return fmt.Errorf("%w: bad type: cannot indirect", ErrRequestFailed)
+	}
+
+	f := r.FieldByName("HTTPResponse")
```

Comment:
Done, sort-of.  The docs mean that _`r`_ has to have a struct type.  It does, so that's OK.

Modified anyway to test for the kind of `r` -- it will be an internal (logic) error reported as an error, but probably less scary than a `panic`.